File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        
        if(!(str1+str2).equals(str2+str1)){
            return "";
        }

        int gcd = gcd(str1.length(), str2.length());
        return str1.substring(0,gcd);




    }

    // eg a = 6 and b = 3
    // gcd(b, a%b) = gcd(3,6%3=0)
    private int gcd(int a, int b){
        // if (b == 0){
        //     return a;
        // }
        // else{
        //     gcd(b, a%b);
        // }
        
        return b == 0 ? a : gcd(b, a%b);

    }
}

File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        
        # first we concatenate both the strings to check whether the result of concatenation is same or not.
        #  if str1 = ABCABC and str2 = ABC, then concatenation should be ABCABCABC if we do (str1+str2 or str2+str1)
         # for str1 = LEET and str2 = CODE
        if str1 + str2 != str2 + str1: 
            return ""

        # now calcuate the gcd of both the string and return the substring starting from 0 to gcd of str1,str2
        from math import gcd
        result = str1[0: gcd(len(str1),len(str2))]

        return result